### ISSUE TITLE
Hangs forever on id="<%= some %>"

### ISSUE THREAD

**Issue by thebravoman** [threedaymonk/htmlbeautifier#51]
I have an html.erb file. 
htmlbeautifier is version (1.3.1)

I have a large erb file. 
I execute 

````
htmlbeautifier the_file.html.erb
````
This command hangs forever and does not return.

In the file I have 

````
<article id='<%= @holiday_content.to_param %>'>
              </article>
````

If I remove the id and leave it to just
````
<article id='ddd'>
              </article>
````

It works again.

**Comment by thebravoman** [comment]
Here is a more complete example

this is the whole file after beautifying
````
<article id="<% %>" class="episode-partial">
                <div class="episode sheet episode-sheet">
                  <a href="dd" class="" >
                    <div class="screenshot">
                      <img alt="Image for <%= @holiday_content.title %>" class="img-thumbnail top" src="<%= @holiday_content.thumbnail.url(:p1080) %>"/>
</div>
</a>
</div>
</article>

````
The file is not formatted and the lines are not indented. 

Note that there is <article id="<% > "...

If we set <article id="dd"> than this happens which

````
<article id="dd" class="episode-partial">
  <div class="episode sheet episode-sheet">
    <a href="dd" class="" >
      <div class="screenshot">
        <img alt="Image for <%= @holiday_content.title %>" class="img-thumbnail top" src="<%= @holiday_content.thumbnail.url(:p1080) %>"/>
      </div>
    </a>
  </div>
</article>
````

**Comment by cgimenez** [comment]
Had the same problem and managed to understand what's happening, but without finding a solution.
This bug is partly related with htmlbeautifier and party with Ruby core libraries.

What happens is that when the htmlbeautifier parser attempts to detect what an html line is, it fails with '<%= @holiday_content.to_param %>', but not everytime. It depends on the complexity of the erb template.

The offending statement in code is at https://github.com/threedaymonk/htmlbeautifier/blob/4f4fa8c293f42e5475370d3bd8595eec2f4171d0/lib/htmlbeautifier/parser.rb#L30

When iterating over the regexp patterns, depending on the pattern and the template content the @scanner.scan(pattern) never returns and push the cpu to 99%. @scanner.scan(pattern) is a call to Ruby core library StringScanner.scan. 

So, is there a bug in the StringScanner.scan code ? Or are htmlbeautifier too complex or buggy ? I dunno...

For me, the offending regexp pattern is 

```
(?m-ix:(<%-?=?)(.*?)(-?%>))
(?m-ix:<!--\[.*?\]>)
(?m-ix:<!\[.*?\]-->)
(?m-ix:<!--.*?-->)
(?m-ix:<!.*?>)
(?mi-x:(<script(?mx-i: (?:<%.*?%>|[^>])* )>)(.*?)(<\/script>))
(?mi-x:(<style(?mx-i: (?:<%.*?%>|[^>])* )>)(.*?)(<\/style>))
(?mi-x:(<pre(?mx-i: (?:<%.*?%>|[^>])* )>)(.*?)(<\/pre>))
(?mi-x:(<textarea(?mx-i: (?:<%.*?%>|[^>])* )>)(.*?)(<\/textarea>))
(?m-ix:<(?mix:
      area | base | br | col | command | embed | hr | img | input | keygen |
      link | meta | param | source | track | wbr
    )(?: (?mx-i: (?:<%.*?%>|[^>])* ))?\/?>)
(?m-ix:<\w+(?: (?mx-i: (?:<%.*?%>|[^>])* ))?\/>)
```
 
But debugging this kind of stuff is a pitta, I'm not a regexp guru ;-)

**Comment by cgimenez** [comment]
Just walking the forks, I found this one that seems to do the job : https://github.com/jeremyf/htmlbeautifier

Thanks you Jeremy !

**Comment by buncis** [comment]
why its not merged yet?

**Comment by amcvitty** [comment]
Switching to this fork https://github.com/jeremyf/htmlbeautifier solved an issue for me too! So +1 👍 on merging the changes there upstream!

@threedaymonk are you open to merging and releasing a 1.3.2 on this? I see it's been a while since you were last working on it. 

My issue: I had an issue with formatting some ERB files that also boils down to the handling of self-closing tags I think. The symptom was that it flattened everything after it failed to handle the `<img />` tag, and that's caused by it failing to parse, as can be seen here by adding stop_on_errors. 

Here's my test to add to spec/behavior_spec.rb
```ruby
  # This is probably a bug. It throws "Extraneous closing tag on line 6"
  # Seems to be related to the interplay between the ERB tags, the script tag and
  # the self-closing HTML void element.
  it "ignores HTML fragments in javascript strings in <script> tags" do
    source = code <<-END
    <div class="<%= get_class %>" ></div>
    <script>
      var warning_message = "<%= confirm_data %>";
      $('#img').html('<img src="/myimg.jpg" />')
      $('#errors').html('<div class="alert alert-danger" role="alert">' + error_message + '</div>')
    </script>
    END

    expect(described_class.beautify(source, stop_on_errors: true)).to eq(source)
  end
```

**Comment by threedaymonk** [comment]
I'm not particularly keen to chase down changes in forks, but If someone wants to submit a PR with the relevant changes, I'll take a look.

**Comment by amcvitty** [comment]
Thanks I’ll chase it down and pull that together then!
________________________________

**Comment by amcvitty** [comment]
@threedaymonk here's my PR: https://github.com/threedaymonk/htmlbeautifier/pull/60

**Comment by threedaymonk** [comment]
This should now be fixed in version 1.4.0.
